tests/Unit/ChangelogTest.php

```diff
@@ -0,0 +1,269 @@
+<?php
+
+declare(strict_types=1);
+
+namespace AsyncAws\Test\Unit;
+
+use PHPUnit\Framework\TestCase;
+use Symfony\Component\Finder\Finder;
+
+/**
+ * Verify CHANGELOG entries are present.
+ */
+class ChangelogTest extends TestCase
+{
+    /**
+     * @dataProvider provideChangedlogFiles
+     */
+    public function testChangelogFormat(string $changelogPath)
+    {
+        $lines = explode("\n", file_get_contents($changelogPath));
+
+        self::assertGreaterThan(3, \count($lines), 'CHANGEOG must not be empty');
+        self::assertSame('# Change Log', $lines[0], 'CHANGELOG MUST start with "# Change Log"');
+        self::assertSame('', $lines[1], 'CHANGELOG MUST have a second line empty');
+        self::assertSame('## NOT RELEASED', $lines[2], 'CHANGELOG MUST have a "## NOT RELEASED" section');
+
+        $inTitle1 = false;
+
+        $inTitle2 = false;
+        $title2Empty = false;
+
+        $inTitle3 = false;
+        $title3Empty = false;
+        $title3s = [];
+
+        $needEmptyLine = false;
+        $wasEmptyLine = false;
+        $finalWords = false;
+        foreach ($lines as $index => $line) {
+            if ($needEmptyLine) {
+                self::assertSame('', $line, 'CHANGELOG line ' . $index . ' MUST be empty');
+                $needEmptyLine = false;
+                $wasEmptyLine = true;
+
+                continue;
+            }
+            self::assertFalse($finalWords, 'CHANGELOG line ' . $index . ' MUST not contains anything after "First version"');
+            if ('' === $line) {
+                $wasEmptyLine = true;
+
+                continue;
+            }
+
+            if (0 === strpos($line, '# ')) {
+                self::assertFalse($inTitle1, 'CHANGELOG line ' . $index . ' MUST contain a single Title');
+                $inTitle1 = true;
+                $needEmptyLine = true;
+                $wasEmptyLine = false;
+
+                continue;
+            }
+
+            if (0 === strpos($line, '## ')) {
+                self::assertTrue($wasEmptyLine, 'CHANGELOG line ' . ($index - 1) . ' MUST be empty');
+                self::assertTrue($inTitle1, 'CHANGELOG line ' . $index . ' MUST have title2 inside a title1');
+                self::assertFalse($title2Empty, 'CHANGELOG line ' . $index . ' MUST not have empty title2');
+                self::assertFalse($title3Empty, 'CHANGELOG line ' . $index . ' MUST not have empty title3');
+                self::assertMatchesRegularExpression('/^## (NOT RELEASED|\d+\.\d+\.\d+)$/', $line, 'CHANGELOG line ' . $index . ' MUST be "## X.Y.Z"');
+                $inTitle2 = true;
+                $inTitle3 = false;
+                $title3s = [];
```

Comment:
Before resetting the list of title 3, should we assert the order of title 3 of the previous section to ensure consistent order.

Reply:
Yeah, I hesitated about this. But fear the diff to fix the existing changelogs...


let's do it once for all